### DISCUSSION TITLE
Existing DSL for JSON Schema?

### DISCUSSION THREAD

**Discussion by chris-hailstorm** [textX/textX#329] [Q&A]
Textx is an outstanding tool, thank you.

Is anyone aware of a JSON Schema implementation in Textx?

**Comment by igordejanovic** [comment]
I'm not aware of any project using textX for JSON Schema implementation.
It shouldn't be hard to define grammar for JSON Schema in textX. You can start with [JSON example](https://github.com/igordejanovic/textX/tree/master/examples/json).
Actually, JSON schema is JSON file so it can be parsed with any JSON parser (even textX JSON example) and validated and analysed afterwards. You could then extract schema validation rules and apply them to subsequent JSON files. JSON schema is meta-language, thus it could be applied to itself.

But, if you want standard JSON Schema implementation there are solution for Python that do that well and probably will parse much faster than textX. textX is perfect for situation where you want to parse something that doesn't have readily available parser, like some custom textual formats or DSLs.
Is there any specific reason not to use existing JSON Schema libs?

**Comment by chris-hailstorm** [comment]
I use other JSON schema tools and they do work well.

However there are some domains that could be expressed better with a
combination of standard JSON schema and custom textx objects -- I was
hoping to combine the two.

Thanks

**Comment by igordejanovic** [comment]
Sounds interesting. Give it a try if you wish and post here if you get stuck along the way.
I'll be happy to help.

**Comment by goto40** [comment]
You can easily reference an arbitrary non textX object (e. g., from a json model) from a textX model. Maybe have a look at https://github.com/igordejanovic/textX/blob/master/tests/functional/test_scoping/test_reference_to_nontextx_attribute.py
